docs/proposals/scheduling/20220530-fine-grained-cpu-orchestration.md

```diff
@@ -0,0 +1,703 @@
+  CPUBindPolicyNone CPUBindPolicy = "None"
+  // CPUBindPolicyFullPCPUs favor cpuset allocation that pack in few physical cores
+  CPUBindPolicyFullPCPUs CPUBindPolicy = "FullPCPUs"
+  // CPUBindPolicyDistributeAcrossPCPUs favor cpuset allocation that evenly allocate logical cpus across physical cores
+  CPUBindPolicyDistributeAcrossPCPUs CPUBindPolicy = "DistributeAcrossPCPUs"
+  // CPUBindPolicyConstrainedBurst constrains the CPU Shared Pool range of the Burstable Pod
+  CPUBindPolicyConstrainedBurst CPUBindPolicy = "ConstrainedBurst"
+)
+
+type CPUExclusivePolicy string
+
+const (
+  // CPUExclusivePolicyNone does not perform any exclusive policy
+  CPUExclusivePolicyNone CPUExclusivePolicy = "None"
+  // CPUExclusivePolicyPCPULevel represents mutual exclusion in the physical core dimension 
+  CPUExclusivePolicyPCPULevel CPUExclusivePolicy = "PCPULevel"
+  // CPUExclusivePolicyNUMANodeLevel indicates mutual exclusion in the NUMA topology dimension
+  CPUExclusivePolicyNUMANodeLevel CPUExclusivePolicy = "NUMANodeLevel"
+)
+```
+
+- The `CPUBindPolicy` defines the CPU binding policy. The specific values are defined as follows:
+   - `CPUBindPolicyNone` or empty value does not perform any bind policy. It is completely determined by the scheduler plugin configuration.
+   - `CPUBindPolicyFullPCPUs` is a bin-packing policy, similar to the `full-pcpus-only=true` option defined by the kubelet, that allocate full physical cores. However, if the number of remaining logical CPUs in the node is sufficient but the number of full physical cores is insufficient, the allocation will continue. This policy can effectively avoid the noisy neighbor problem.
+   - `CPUBindPolicyDistributeAcrossPCPUs` is a spread policy. If the node enabled Hyper-Threading, when this policy is adopted, the scheduler will evenly allocate logical CPUs across physical cores. For example, the current node has 8 physical cores and 16 logical CPUs. When a Pod requires 8 logical CPUs and the `CPUBindPolicyDistributeAcrossPCPUs` policy is adopted, the scheduler will allocate an logical CPU from each physical core. This policy is mainly used by some latency-sensitive applications with multiple different peak-to-valley characteristics. It can not only allow the application to fully use the CPU at certain times, but will not be disturbed by the application on the same physical core. So the noisy neighbor problem may arise when using this policy.
+   - `CPUBindPolicyConstrainedBurst` a special policy that mainly helps K8s Burstable/Koordinator LS Pod get better performance. When using the policy, the scheduler uses the Pod's Limit as a resource request. Priority is given to selecting the node whose NUMA node satisfies the Limit and the sum of the Limit among the NUMA nodes is the smallest. After the scheduling is successful, the scheduler will update `scheduler.koordinator.sh/resource-status` in the Pod, declaring the CPU Shared Pool to be bound.
```

Comment:
Do we want this cpuset to be computed on the scheduler or on the koordlet?

Reply:
koord-scheduler is responsible for filtering out nodes with suitable CPU Shared Pool by NUMA Node. koordlet is responsible for computing and binding CPUSet.